Undo of weightsystem editing

bstoeger opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [x] Functional change
- [ ] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
We have to progress on that front - this implements undo of the weight system actions (add, delete, edit). That's the easy part of the equipment tab, but it gave me some ideas on how to handle the much harder part: cylinders. The problem with cylinders is that they are used in two places: normal editing and planner. That will be fun.

Overall, this was a bit harder than expected, because when hovering over the weightsystem-types, the old code would overwrite the displayed dive. The logic was changed so that the model now has a "temporary" weightsystem line that it can display. This can be set, committed or reset.

I don't like how the model calls the undo-command, but that's the easiest way for now.

This is only lightly tested.

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->
Done.
### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->
Unsure.

dirkhh commented:
Hmm - I'm having a weird issue.
Dive without weight.
I add weight called "integrated" it autocompletes to "3.6kg" (I apparently used that name before).
I try to update this to 4kg
It refuses to update, switches back to 3.6kg
I edit other dives
I try again. still can't change the amount for that weight

bstoeger commented:
Thanks for testing - I will look into that.

Off-topic: I started with the linear-model thing, but was side-tracked by other things (e.g. reading a thesis). Then lost interest when I realized that in QML we use roles, on desktop columns. Will try to do more today.

bstoeger commented:
> I think it would be cleaner to just have conditional code there that uses different strings for n=1 and n>1 -- relying on the translation mechanism to take care of that seems a bit excessive.

Fine with me - but note that last time we had an if, you said that we use the translation thing for that. In that case though it was only the plural form that changed.

But perhaps we should even leave in the "weight edited (1 dive)"? Up to you.

dirkhh commented:
> Off-topic: I started with the linear-model thing, but was side-tracked by other things (e.g. reading a thesis). Then lost interest when I realized that in QML we use roles, on desktop columns. Will try to do more today.

Ahh, missed that comment earlier - see my email to you :-)
Yes, roles vs. columns was one of the things I noticed as well, but that seemed like a rather straight forward translation (famous last words). I'm quite eager to see this - knowledge in my brain appears to work like OS paging algorithms... so right now the understanding of the QML dive list is "paged in" and I have worked on understanding the models and would love to put this "warm cache" situation to good use... before something else displaces it again and I need to start reading everything again...

dirkhh commented:
> > I think it would be cleaner to just have conditional code there that uses different strings for n=1 and n>1 -- relying on the translation mechanism to take care of that seems a bit excessive.
> 
> Fine with me - but note that last time we had an if, you said that we use the translation thing for that. In that case though it was only the plural form that changed.
> 
> But perhaps we should even leave in the "weight edited (1 dive)"? Up to you.

Ahh. Yeah - this is subtle. If all we do is "%n dive(s)" then using the plural algorithm in the translation is really easiest. The situation here seemed different in that you wanted no text when n is 1. Because we still need the plural thing for the n>1 case (as some languages have a different form for 2 than for more than 2). But for the n=1 case we just want to drop part of the text and that's harder to get right for the translators. It's doable - but I think it would be easier to separate that out in code.

I realize writing this that maybe I'm not fully consistent in my reasoning here :-)

bstoeger commented:
> Hmm - I'm having a weird issue.

I can reproduce this and it's an UI issue. The crucial point is that when you choose a different weight type, you don't click on it, but move away with the mouse. By explicitly clicking on a (different) weight-type it will work again.

The problem is, if the user doesn't click on an item, we don't get the correct closeEditor signal that says "commit this". This is annoying, because if we react to every closeEditor signal, we get two signals when the user clicks on an item. I will have to think about that.

bstoeger commented:
I hope I fixed it, but this whole delegate stuff is very obscure and I'm not sure how stable it is across Qt versions. We might have to adjust the Hint-switch statement.

bstoeger commented:
Argh. Stopped working again - will check that out tomorrow. We're getting a spurious setModelData call after having committed the change.

bstoeger commented:
Alright, back to this...

Rebased on current master (surprisingly painful) and changed to weightsystem-model logic. I couldn't get the editor-hint and setValue calls under control (too many different cases: clicked, tab-pressed, esc-pressed, etc.). Therefore, I chose a different approach: if the item is set to the already existing item, we exit (or don't enter) temporary-mode.

@dirkhh: please test if you find time.

dirkhh commented:
And now this one conflicts as well.
I need to be faster merging things so I don't end up causing all this bit rot to add extra cycles.
:-(

dirkhh commented:
I merged this locally and the resulting build behaves strangely.
Simple test file with a could of dives with no weights.
Click the plus to add a weight system.
No visual change.
Click the plus again.
Nothing.
Click the plus for a tank. Ahh, that works. Discard.
Now the two empty weigh systems show up...

Second attempt, same file
add weight
nothing
switch to different dive
try to switch back
app hangs.

bstoeger commented:
Interesting. Doesn't happen for me (with unmerged PR).

bstoeger commented:
Oh - found a way to reproduce this - you mustn't change dive first. Appears to be an initialization thing.

bstoeger commented:
Note that even if you *do* have weight and cylinder - before you switch dive they aren't shown.

bstoeger commented:
This particular problem should be fixed by #2400 - let's get that in first and then rebase.

bstoeger commented:
> I merged this locally and the resulting build behaves strangely.

Rebased on master and that particular problem should be fixed.

bstoeger commented:
@dirkhh: Added a commit to make the leave-out parentheses from text for 1 dives explicit.

I did this as a separate commit because we already used the pattern elsewhere and so we stay consistent.

I'm not 100% convinced that we should leave out the parentheses for single-dive edits and could also simply remove the translator comments - feel free to leave out the last commit.
